src/core/expr/head_func_qcut.cc

```diff
@@ -0,0 +1,196 @@
+
+
+ptrHead Head_Func_Qcut::make(Op, const py::otuple& params) {
+  xassert(params.size() == 1);
+  return ptrHead(new Head_Func_Qcut(params[0]));
+}
+
+
+
+Workframe Head_Func_Qcut::evaluate_n(
+    const vecExpr& args, EvalContext& ctx, bool) const
+{
+
+  if (ctx.has_groupby()) {
+    throw NotImplError() << "qcut() cannot be used in a groupby context";
+  }
+
+  int32_t nquantiles_default = 10;
+  Workframe wf = args[0].evaluate_n(ctx);
+  const size_t ncols = wf.ncols();
+
+  int32vec nquantiles(ncols);
+  bool defined_nquantiles = !py_nquantiles_.is_none();
+  bool nquantiles_list_or_tuple = py_nquantiles_.is_list_or_tuple();
+
+  if (nquantiles_list_or_tuple) {
+    py::oiter py_nquantiles = py_nquantiles_.to_oiter();
+    if (py_nquantiles.size() != ncols) {
+      throw ValueError() << "When `nquantiles` is a list or a tuple, its length must be "
+        << "the same as the number of columns in the frame/expression, i.e. `"
+        << ncols << "`, instead got: `" << py_nquantiles.size() << "`";
+
+    }
+
+    size_t i = 0;
+    for (auto py_nquantile : py_nquantiles) {
+      int32_t nquantile = py_nquantile.to_int32_strict();
+      if (nquantile <= 0) {
+        throw ValueError() << "All elements in `nquantiles` must be positive, "
+                              "got `nquantiles[" << i << "`]: `" << nquantile << "`";
+      }
+
+      nquantiles[i++] = nquantile;
+    }
+    xassert(i == ncols);
+
+  } else {
+    if (defined_nquantiles) {
+      nquantiles_default = py_nquantiles_.to_int32_strict();
+      if (nquantiles_default <= 0) {
+        throw ValueError() << "Number of quantiles must be positive, "
+                              "instead got: `" << nquantiles_default << "`";
+      }
+    }
+
+    for (size_t i = 0; i < ncols; ++i) {
+      nquantiles[i] = nquantiles_default;
+    }
+  }
+
+  // Qcut workframe in-place
+  for (size_t i = 0; i < ncols; ++i) {
+    Column coli = wf.retrieve_column(i);
+
+    if (coli.ltype() == dt::LType::STRING || coli.ltype() == dt::LType::OBJECT)
+    {
+      throw TypeError() << "`qcut()` cannot be applied to "
+        << "string or object columns, instead column `" << i
+        << "` has an stype: `" << coli.stype() << "`";
+    }
+
+    coli = Column(new Latent_ColumnImpl(new Qcut_ColumnImpl(
+             std::move(coli), nquantiles[i]
+           )));
+    wf.replace_column(i, std::move(coli));
+  }
+
+  return wf;
+}
+
+}}  // dt::expr
+
+
+namespace py {
+
+static oobj make_pyexpr(dt::expr::Op opcode, otuple targs, otuple tparams) {
+  size_t op = static_cast<size_t>(opcode);
+  return robj(Expr_Type).call({ oint(op), targs, tparams });
+}
+
+
+static const char* doc_qcut =
+R"(qcut(cols, nquantiles=10)
+--
+
+Bin all the columns in a Frame/f-expression into equal-population
+discrete intervals, i.e. quantiles. In reality, for some data
+these quantiles may not have exactly the same population.
+
+Parameters
+----------
+cols: Frame | f-expression
```

Comment:
I guess this should be saying just `f-expression` now, with the rest of the doc updated accordingly

Reply:
Why? This is an input parameter, that can also be a `Frame`.

Reply:
I guess this one should be changed to `Expr` instead

```
return: Frame | Expr
    Frame/f-expression, where each column is filled with
    the respective quantile ids.
```

Reply:
The argument can actually be either an FExpr, or Frame, or pandas DataFrame/Series, or numpy array, or a python range, or a simple number, etc. In reality we will take the argument and attempt to convert into an FExpr, which works for many different types of values (and the set of types is going to only expand in the future). The resulting FExpr must also implement the `evaluate_n()` method, and most of them do but some don't (like `Head_Literal_Type`).

Reply:
So, the CORRECT way would be to come up with a new term (like `FExpr-like` or `FExpr-able` or I don't know), and then create a separate help page to describe what that term actually means, then make it so that this type would automatically link to its corresponding help page in Sphinx.

A shorter way would be to simply write the new term here, and then leave it unexplained for now, creating a new ticket for finishing the documentation part.

Reply:
Don't we already have any function that support the same input type?